Rsite/source/tutorial/CustomLossGluon.md

```diff
@@ -0,0 +1,235 @@
+
+However, we may sometimes want to solve problems that require customized loss functions and this tutorial shows how we can do that in Gluon. We will show it at the example of contrastive loss which is typically used in Siamese networks.
+
+```python
+import matplotlib.pyplot as plt
+import mxnet as mx
+from mxnet import autograd, gluon, nd
+from mxnet.gluon.loss import Loss
+import random
+```
+
+### What is contrastive Loss
+
+[Contrastive loss](http://yann.lecun.com/exdb/publis/pdf/hadsell-chopra-lecun-06.pdf) is a distance-based loss function. During training pairs of images are fed into a model. If the images are similar the loss function will return 0, otherwise 1. 
+
+<img src="contrastive_loss.jpeg" width="400">
+
+*Y* is a binary label indicating similarity between training images. Contrastive loss uses the Euclidean distance *D* between images. The contrastive loss is the sum of 2 terms: 
+ - the first one indicates the loss for a pair of similar points
+ - the second one indicates the loss for a pair of dissimilar points
+
+The loss function uses a margin *m* which is has the effect that dissimlar pairs only contribute if their loss is within a certain margin. 
+
+In order to implement such a customized loss function in Gluon, we only need to define a new class that is inhertiting from the Loss base class. We then define the contrastive loss in the ```hybrid_forward```. This function takes the images ```image1```, ```image2``` and the label which defines whether  ```image1``` and ```image2``` are similar (=0) or  dissimilar (=1). The input F is either an ```mxnet.ndarry``` or an ```mxnet.symbol``` if we hybridize the network. Gluon's Loss base class is in fact a HybridBlock. This means we can either run fully imperatively or symbolically. When we hybridize our custom loss function, we can get performance speedups.
+
+
+```python
+class ContrastiveLoss(Loss):
+    
+    def __init__(self, margin=6., weight=None, batch_axis=0, **kwargs):
+        
+        super(ContrastiveLoss, self).__init__(weight, batch_axis, **kwargs)
+        self.margin = margin
+        
+    def hybrid_forward(self, F, image1, image2, label):
+        
+        distances           = image1 - image2
+        distances_squared   = F.sum(F.square(distances), 1, keepdims=True)
+        euclidean_distances = F.sqrt(distances_squared + 0.0001)
+        d = F.clip(self.margin - euclidean_distances, 0, self.margin)
+        loss = (1 - label) * distances_squared +  label * F.square(d)
+        loss = 0.5*loss
+        
+        return loss
+    
+loss = ContrastiveLoss(margin=6.0)
+```
+
+### Define the Siamese network
+A Siamese network consists of 2 identical networks, that share the same weights. They are trained on pair of images and each network processes one image. The label defines whether the pair of images is similar or not. The Siamese network learns to differentiate between two input images. 
```

Comment:
Would you be able to add a link to Siamese networks - a source for more information?